File: 2026-04-April-GFG-POTD/April 25 - Opposite Sign Pair Reduction/Solution.py
class Solution:
    def reducePairs(self, arr):
        stack = []

        for num in arr:
            add = True

            while stack:
                top = stack[-1]

                # Opposite signs
                if (top > 0 and num < 0) or (top < 0 and num > 0):
                    stack.pop()

                    if abs(top) > abs(num):
                        num = top
                    elif abs(top) < abs(num):
                        pass
                    else:
                        add = False
                        break
                else:
                    break

            if add:
                stack.append(num)

        return stack

File: 2026-04-April-GFG-POTD/April 25 - Opposite Sign Pair Reduction/Solution.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> reducePairs(vector<int>& arr) {
        stack<int> st;

        for (int num : arr) {
            bool add = true;

            while (!st.empty()) {
                int top = st.top();

                // Opposite signs
                if ((top > 0 && num < 0) || (top < 0 && num > 0)) {
                    st.pop();

                    if (abs(top) > abs(num)) {
                        num = top; // keep top
                    } 
                    else if (abs(top) < abs(num)) {
                        // keep num
                    } 
                    else {
                        add = false; // both removed
                        break;
                    }
                } else {
                    break;
                }
            }

            if (add) {
                st.push(num);
            }
        }

        vector<int> result;
        while (!st.empty()) {
            result.push_back(st.top());
            st.pop();
        }

        reverse(result.begin(), result.end());
        return result;
    }
};
